### ISSUE TITLE
Spec

### ISSUE THREAD

**Issue by stevepeak** [asyncy/lsp#1]
> https://langserver.org

## Features
- [Code completion](https://microsoft.github.io/language-server-protocol/specification#completion-request-leftwards_arrow_with_hook)
  - This should be completed through a plugin interface. Since Storyscript is an open language Asyncy will have it's own plugin for auto-completing according from the Asyncy Hub
- [Hover](https://microsoft.github.io/language-server-protocol/specification#hover-request-leftwards_arrow_with_hook)
- Jump to def
- [Workspace symbols](https://microsoft.github.io/language-server-protocol/specification#workspace-symbols-request-leftwards_arrow_with_hook)
- Find references
- [Diagnostics](https://microsoft.github.io/language-server-protocol/specification#diagnostic)
  - Syntax
  - Type Checking
  - Undefined Variable
  - Formatting 

# Examples
- Ballerina in Java https://github.com/ballerina-platform/ballerina-lang/tree/master/language-server
- https://github.com/sourcegraph/python-langserver [stage:alpha]
- https://github.com/palantir/python-language-server

**Comment by Vesuvium** [comment]
Ideally, the lsp will rely on two separate tools, one for autocomplete and
one for diagnostics. They would share some code between them, but they are
serving different purposes so it makes more sense to have them separated.

## Code completion

There are three steps required for code completion:

- Parse Storyscript files for the project
- Generate constructs for guesses
- Make guesses based on available constructs and current cursor position

It might look simple, but these steps can be very complex, depending on
how precise, user-friendly and sophisticated they must be.

There is also a speed requirement as this must happens smoothly enough for
comfortable typing. It does not seem there are common-known problems with the
speed, so it should be ok.

Hovering, jumping, symbols, references can be in the code completion. They could
be separate but I don't see the reason for it.

## Diagnostics

Diagnostics would need to identify weak points of the code and check them:

- Parse Storyscript files for the project
- Find weak points
    - Check for syntax errors
    - Check all vars are defined
    - Check the return types of functions
- Formatting is different and needs to manipulate bits of code and transform them
- Refactoring is similar for some ways to formatting, transform code in some way.

## Editor-specific plugins

Editors might require specific packages that act as wrappers; for example
Atom requires a small wrapper to start the LSP and send and route messages from
editor to lsp.

The python one it's 200 Javascript lines.

## Crunchy, tasty numbers

It's hard to get an idea of how much work this will need, not
having direct experience in this domain and given that most of the tools around
do not have a clear separation of responsibilities, doing more than they
are supposed to do.

Similar packages don't really help, because they are for different languages.
Python has a more intricate syntax, but is not as dynamic as Storyscript.

- Python's LSP has about 3k lines of code.
- Jedi sits at about 15k lines and it comes with lots of features. Simpler
  autocompletion for Node or Elixir takes about 3k lines of code.

Jedi also relies on a custom Python parser (parso) to show more than one syntax error
in the same file and to support different versions of Python.
Lark does not support multiple errors in the same source, but it's possible to add it.

**Comment by judepereira** [comment]
That's a pretty solid breakdown. Yep, we can split this into two parts:

1. auto complete
2. diagnostics

@Vesuvium Do you know how complex refactoring, such as renaming a symbol works? If all the symbols are indexed in the IDE (provided by our lsp), will it (the IDE) take care of it? Or does our lsp edit files on the fly?

**Comment by Vesuvium** [comment]
It looks like it's a string replacement. The LSP just tells the IDE to delete the old part and write the new one.

**Comment by judepereira** [comment]
That’s interesting, how does this work then? For every request for completion or anything, will the LSP module read the full storyscript file at all times?

**Comment by judepereira** [comment]
Will the LSP module keep reloading the file from based on something like inotify? (Given that the request for completion or go to symbol request only includes the cursor location)

**Comment by Vesuvium** [comment]
Yes. That's why at some point there will be caching to avoid compiling again unchaged files. The IDE sends requests to the lsp at leisure.

**Comment by judepereira** [comment]
Woah, okay. We’ll have to see how this goes in terms of performance.

**Comment by wilzbach** [comment]
> Lark does not support multiple errors in the same source, but it's possible to add it.

How would that work? The only way I see to support this is to write our own parser.
See also e.g. https://github.com/storyscript/storyscript/issues/457
